### DISCUSSION TITLE
Liquid Tag to generate a table of contents

### DISCUSSION THREAD

**Discussion by jmfayard** [forem/forem#19932] [Feature Requests]
My article [ 20+ Lessons I've Learned Writing on DEV for 4 Years ](https://dev.to/jmfayard/20-lessons-i-ve-learned-writing-on-dev-for-4-years-4nk3) is quite long
But it does have an handy table of contents
![2023-08-13 toc](https://github.com/forem/forem/assets/459464/62f5e9f1-a90f-4635-bb6d-7e7730a74752)

"_Here I want a table of contents_" is a feature you can have litterally everywhere
- google docs
- latex with `/listoffigures`, `/listoftables` and friends
- [Asciidoc automatic table of contents](https://docs.asciidoctor.org/asciidoc/latest/toc/) with just `:toc:`

But crucially it's not present in Markdown. 
But it should really.

There are multiple markdown table of contents generator, for visual studio code, as a web tool.

But I had to try multiple ones to find one that works reliably on DEV.to including when I use smileys and stuff.
It works with just `[TOC]`
https://derlin.github.io/bitdowntoc/
![2023-08-13 TOC 2](https://github.com/forem/forem/assets/459464/1ffe5a6c-5933-414e-96cb-167f4844faea)

**Suggestion**

Add an official liquid tag on Forem `{% toc %}` that would do the same as bitdowntoc
https://developers.forem.com/frontend/liquid-tags

**Alternative:**
https://github.com/forem/forem/discussions/15627 is about having a navigation widget outside of the article.
This is a good idea but I don't think it's an either/or, we could have both.

**Comment by bcouetil** [comment]
I second that request, please 🙏

GitLab flavored Markdown [has it](https://docs.gitlab.com/ee/user/markdown.html#table-of-contents) :

> A table of contents is an unordered list that links to subheadings in the document. You can add a table of contents to issues and merge requests, but you can’t add one to notes or comments. Add either the `[[_TOC_]]` or `[TOC]` tag on its own line.

For now I use https://derlin.github.io/bitdowntoc/
